subxt/src/rpc/rpc.rs

```diff
@@ -569,6 +576,97 @@ impl<T: Config> Rpc<T> {
         Ok(subscription)
     }
 
+    /// Subscribe to finalized blocks and ensure that all missing blocks are filled.
+    ///
+    /// The Substrate's RPC does not guarantee that all finalized blocks are provided.
+    /// This function ensures that all missed blocks are fetched and returned in order.
+    ///
+    /// # Example
+    ///
+    /// The following is a possible scenario where Substrate does not generate an event
+    /// for the second block.
+    ///
+    /// ```sh
+    /// substrate      | Block A | --- | Block B | --- | Block C |
+    /// client         | Block A | ---             --- | Block C |
+    /// ```
+    ///
+    /// This function mitigates this by filling the gaps of missing blocks.
+    ///
+    /// ```sh
+    /// substrate      | Block A | --- | Block B | --- | Block C |
+    /// client         | Block A | --- | Block B | --- | Block C |
+    ///                                ^^^ filled
+    /// ```
+    pub async fn subscribe_finalized_blocks_filled(
+        &self,
+    ) -> Result<impl Stream<Item = Result<T::Header, Error>> + Send + '_, Error> {
+        // Fetch the last finalised block details immediately, so that we'll get
+        // all blocks after this one.
+        let last_finalized_block_hash = self.finalized_head().await?;
+        let last_finalized_block_num = self
+            .header(Some(last_finalized_block_hash))
+            .await?
+            .map(|h| (*h.number()).into());
+
+        let sub = self.subscribe_finalized_blocks().await?;
+
+        // Adjust the subscription stream to fill in any missing blocks.
+        Ok(Rpc::subscribe_to_block_headers_filling_in_gaps(
+            self.clone(),
+            last_finalized_block_num,
+            sub,
+        ))
+    }
```

Comment:
I def like the idea of this, but I'm not suepr fond of _where_ it is; it's more than just an RPC call at this point, which is reall what this RPC module is all about. Eventually we'll also be transitioning to the new RPCs too. For these reasons I don't think we should add more stuff here if we can avoid it.

We have talked previously about a `api.blocks()` API to go alongside `api.constants()`, `api.events()` and so on. Perhaps this is the perfect time to add something there?

So just like we currently have `api.events().subscribe_finalized()`, we can have an `api.blocks().subscribe_finalized()` which will be this guy.

For now we don't have to add any other functionality to this new `BlocksClient` but it'll be a good place to start building stuff much like the EventsClient (eg `api.blocks().at()` to get some Block details, etc).

What do you think?

Reply:
Sounds like a plan! It helps us build the foundation for fetching blocks that we can then reutilize all over the place, like events! Thanks for the suggestion!